src/platform/PICO/bus_quadspi_pico.c

```diff
@@ -0,0 +1,371 @@
+}
+
+// Default timeout for QSPI direct IO operations
+#define QSPI_TIMEOUT_MS 10
+
+// Number of dummy clock cycles per consumed byte in direct-mode loops.
+// Justification: QSPI datasheets specify dummy cycles in bits. Our direct-mode
+// helper consumes dummy time by shifting out whole bytes (8 bits at a time).
+// We therefore convert from bits to bytes using 8 bits/byte. If a controller
+// revision were to support sub-byte dummy handling, this could be revisited.
+#define QSPI_DUMMY_BITS_PER_BYTE 8
+
+static inline bool qmi_direct_io(const uint8_t *tx, uint8_t *rx, size_t count, timeMs_t timeout)
+{
+    size_t tx_remaining = count;
+    size_t rx_remaining = count;
+    // Prevent hangs: break out after timeout in ms
+    const timeMs_t start_ms = millis();
+    while (tx_remaining || rx_remaining) {
+        uint32_t flags = qmi_hw->direct_csr;
+        bool can_put = !(flags & QMI_DIRECT_CSR_TXFULL_BITS);
+        bool can_get = !(flags & QMI_DIRECT_CSR_RXEMPTY_BITS);
+
+        if (can_put && tx_remaining) {
+            qmi_hw->direct_tx = tx ? *tx++ : 0x00;
+            --tx_remaining;
+        }
+        if (can_get && rx_remaining) {
+            uint8_t b = (uint8_t)qmi_hw->direct_rx;
+            if (rx) {
+                *rx++ = b;
+            }
+            --rx_remaining;
+        }
+        if (cmpTimeMs(millis(), start_ms) > (int32_t)timeout) {
+            return false; // timeout; caller will deassert CS and disable direct mode
+        }
+    }
+    return true;
+}
+
+bool quadSpiTransmit1LINE(QUADSPI_TypeDef *instance, uint8_t instruction, uint8_t dummyCycles, const uint8_t *out, int length)
+{
+    UNUSED(instance);
+    uint8_t txbuf[1 + 0];
+    txbuf[0] = instruction;
+
+    pico_qspi_enter_cmd_mode();
+    qmi_direct_enable();
+    qmi_cs1_assert(true);
+    // Send instruction
+    if (!qmi_direct_io(txbuf, NULL, 1, QSPI_TIMEOUT_MS)) {
+        qmi_timeout_cleanup();
+        (void)dummyCycles;
+        return false;
+    }
+    // Send payload (no readback)
+    if (out && length > 0) {
+        if (!qmi_direct_io(out, NULL, (size_t)length, QSPI_TIMEOUT_MS)) {
+            qmi_timeout_cleanup();
+            (void)dummyCycles;
+            return false;
+        }
+    }
+    qmi_cs1_assert(false);
+    qmi_direct_disable();
+    pico_qspi_exit_cmd_mode();
+    (void)dummyCycles;
+    return true;
+}
+
+bool quadSpiReceive1LINE(QUADSPI_TypeDef *instance, uint8_t instruction, uint8_t dummyCycles, uint8_t *in, int length)
+{
+    UNUSED(instance);
+    uint8_t txbuf[1];
+    txbuf[0] = instruction;
+
+    pico_qspi_enter_cmd_mode();
+    qmi_direct_enable();
+    qmi_cs1_assert(true);
+    // Send instruction
+    if (!qmi_direct_io(txbuf, NULL, 1, QSPI_TIMEOUT_MS)) {
+        qmi_timeout_cleanup();
+        return false;
+    }
+    // Consume dummy cycles as bytes
+    while (dummyCycles >= QSPI_DUMMY_BITS_PER_BYTE) {
+        uint8_t z = 0x00;
+        if (!qmi_direct_io(&z, NULL, 1, QSPI_TIMEOUT_MS)) {
```

Comment:
```suggestion
        if (!qmi_direct_io(NULL, NULL, 1, QSPI_TIMEOUT_MS)) {
```

And you can use `qmi_direct_io(NULL, NULL, (dummyCycles  + QSPI_DUMMY_BITS_PER_BYTE - 1) / QSPI_DUMMY_BITS_PER_BYTE , QSPI_TIMEOUT_MS)`